server/src/main/java/org/opensearch/indices/recovery/RecoveryTarget.java

```diff
@@ -219,7 +220,15 @@ public void prepareForTranslogOperations(int totalTranslogOps, ActionListener<Vo
 
     @Override
     public void finalizeRecovery(final long globalCheckpoint, final long trimAboveSeqNo, ActionListener<Void> listener) {
-        ActionListener.completeWith(listener, () -> {
+        final StepListener<Void> segrepListener = new StepListener<>();
+        // Force segment refresh on primary relocation which is recovered as replica initially & reset engine
+        if (this.indexShard.indexSettings().isSegRepEnabled() && this.state().getPrimary() == true) {
+            logger.info("--> Force segment replication event on new primary and switch to ReadWrite (Internal) engine");
+            this.indexShard.performSegmentReplicationRefresh(segrepListener);
```

Comment:
I think we should look at a solution here that doesn't couple IndexShard to our SegmentReplicationTargetService.

What if we added a new req handler inside of SegmentReplicationTargetService that could be invoked from the primary before this step?  It would be similar to a publischeckpoint, but only route to a single shard.

Reply:
+1 - we could pass a consumer (something like SegmentReplicationTargetService:: performSegmentReplicationRefresh) to the recovery process and decouple index shard and the segrep service.